File: examples/python/compliance_check.py
#!/usr/bin/env python3
"""Run a sample compliance verification for window thermal transmittance."""

from __future__ import annotations

import asyncio
import os
import sys
from typing import Any

import httpx

DEFAULT_API_URL = "https://api.normatia.com"


def get_client_config() -> tuple[str, dict[str, str]]:
    api_key = os.getenv("NORMATIA_API_KEY")
    if not api_key:
        raise RuntimeError(
            "Missing NORMATIA_API_KEY environment variable. "
            "Set it to your bearer key, for example: sk-normatia-xxxxx"
        )

    base_url = os.getenv("NORMATIA_API_URL", DEFAULT_API_URL).rstrip("/")
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Accept": "application/json",
        "Content-Type": "application/json",
    }
    return base_url, headers


def print_sources(sources: Any) -> None:
    print("sources:")
    if not isinstance(sources, list) or not sources:
        print("  none")
        return

    for source in sources:
        if isinstance(source, dict):
            title = source.get("title") or source.get("name") or "untitled"
            ref = source.get("reference") or source.get("url") or source.get("id") or "no-reference"
            print(f"  - {title} ({ref})")
        else:
            print(f"  - {source}")


def print_result(payload: dict[str, Any]) -> None:
    print("Compliance Check: Window U-Value")
    print("===============================")
    print("geo_id: ES-41091")
    print("scope: CTE DB-HE")
    print("input: U = 2.7 W/m2K")
    print()
    print(f"is_compliant: {payload.get('is_compliant')}")
    print(f"result_status: {payload.get('result_status')}")
    print(f"limit_value: {payload.get('limit_value')}")
    print(f"unit: {payload.get('unit')}")
    print(f"reasoning: {payload.get('reasoning')}")
    print_sources(payload.get("sources"))


async def main() -> None:
    base_url, headers = get_client_config()
    request_payload = {
        "element": "Ventana exterior con carpintería de aluminio con rotura de puente térmico",
        "parameter": "Transmitancia térmica (valor U)",
        "value": 2.7,
        "unit": "W/m2K",
        "geo_id": "ES-41091",
    }

    try:
        async with httpx.AsyncClient(base_url=base_url, headers=headers, timeout=20.0) as client:
            response = await client.post("/api/v1/verify", json=request_payload)
            response.raise_for_status()
            payload = response.json()
            if not isinstance(payload, dict):
                raise RuntimeError("Unexpected response shape for compliance verification.")

            print_result(payload)

    except httpx.HTTPStatusError as exc:
        print("\nRequest failed with a non-success HTTP status.")
        print(f"Status: {exc.response.status_code}")
        print(f"URL: {exc.request.url}")
        print(f"Body: {exc.response.text}")
        sys.exit(1)
    except httpx.RequestError as exc:
        print("\nNetwork error while calling Normatia API.")
        print(f"Error: {exc}")
        sys.exit(1)
    except RuntimeError as exc:
        print(f"\nConfiguration or response error: {exc}")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: examples/typescript/compliance-check.ts
/**
 * Demonstrates Normatia Compliance Verification endpoint usage with raw fetch.
 */

const DEFAULT_API_URL = "https://api.normatia.com";

const REQUEST_BODY = {
  element: "Ventana exterior con carpintería de aluminio con rotura de puente térmico",
  parameter: "Transmitancia térmica (valor U)",
  value: 2.3,
  unit: "W/m2K",
  geo_id: "ES-41091",
  codes: [{ slug: "cte-db-he", version: "2022" }],
  context:
    "Edificio residencial plurifamiliar en Sevilla, proyecto de rehabilitación, mejora de envolvente exterior.",
};

function getClientConfig(): { apiKey: string; baseUrl: string } {
  const apiKey = process.env.NORMATIA_API_KEY;
  if (!apiKey) {
    console.error("Error: NORMATIA_API_KEY is not set.");
    console.error("Set it first: export NORMATIA_API_KEY='sk-normatia-xxxxx'");
    process.exit(1);
  }

  return {
    apiKey,
    baseUrl: (process.env.NORMATIA_API_URL ?? DEFAULT_API_URL).replace(/\/+$/, ""),
  };
}

function asObject(value: unknown): Record<string, unknown> {
  if (value && typeof value === "object") {
    return value as Record<string, unknown>;
  }
  return {};
}

function asString(value: unknown): string | undefined {
  return typeof value === "string" ? value : undefined;
}

function formatOutput(value: unknown): string {
  if (typeof value === "string") {
    return value;
  }
  return JSON.stringify(value, null, 2);
}

function formatValue(value: unknown): string {
  if (typeof value === "number" || typeof value === "boolean") {
    return String(value);
  }
  if (typeof value === "string") {
    return value;
  }
  if (value === null || value === undefined) {
    return "n/a";
  }
  return JSON.stringify(value);
}

async function postJson(
  config: { apiKey: string; baseUrl: string },
  path: string,
  body: unknown,
): Promise<unknown> {
  const url = `${config.baseUrl}${path}`;

  let response: Response;
  try {
    response = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${config.apiKey}`,
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
  } catch (error) {
    throw new Error(`Network error while calling ${url}: ${String(error)}`);
  }

  const raw = await response.text();
  let payload: unknown = null;
  if (raw) {
    try {
      payload = JSON.parse(raw);
    } catch {
      payload = raw;
    }
  }

  if (!response.ok) {
    throw new Error(`Request failed (${response.status} ${response.statusText}) for ${url}\n${formatOutput(payload)}`);
  }

  return payload;
}

function printSources(sources: unknown): void {
  console.log("\nSources");
  console.log("=======");

  if (!Array.isArray(sources) || sources.length === 0) {
    console.log("No sources returned.");
    return;
  }

  for (const [index, item] of sources.entries()) {
    const source = asObject(item);
    const title = asString(source.section_title) ?? asString(source.title) ?? "untitled section";
    const reference = asString(source.reference) ?? asString(source.url) ?? "no-reference";
    console.log(`${index + 1}. ${title} | ${reference}`);
  }
}

async function main(): Promise<void> {
  const config = getClientConfig();

  console.log("\n== Compliance check: Window U-value in Seville (ES-41091) against CTE DB-HE ==");

  const payload = asObject(await postJson(config, "/api/v1/verify", REQUEST_BODY));

  const isCompliant = payload.is_compliant;
  const resultStatus =
    asString(payload.result_status) ?? (typeof isCompliant === "boolean" ? (isCompliant ? "compliant" : "non-compliant") : "unknown");

  const providedValue = payload.provided_value ?? REQUEST_BODY.value;
  const providedUnit = asString(payload.unit) ?? REQUEST_BODY.unit;
  const limitValue = payload.limit_value ?? payload.requirement ?? "n/a";

  console.log("\nCompliance status");
  console.log("=================");
  console.log(`is_compliant: ${formatValue(isCompliant)}`);
  console.log(`result_status: ${resultStatus}`);

  console.log("\nProvided vs limit");
  console.log("=================");
  console.log(`provided: ${formatValue(providedValue)} ${providedUnit}`);
  console.log(`limit: ${formatValue(limitValue)} ${providedUnit}`);

  const reasoning = asString(payload.reasoning) ?? asString(payload.explanation);
  if (reasoning) {
    console.log("\nReasoning");
    console.log("=========");
    console.log(reasoning);
  }

  printSources(payload.sources);
}

main().catch((error) => {
  console.error("\nCompliance check example failed.");
  console.error(error instanceof Error ? error.message : String(error));
  process.exit(1);
});
